torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -60,89 +65,155 @@ def reset_table(self) -> None:
+# Tracing is done by replacing the input pointers with Proxy objects that
+# track mutation. Each triton language function is monkey patched to
+# either detect the mutation or return a fresh scalar object.
+def identify_mutated_tensors(kernel, kwargs):
+    from triton import language as tl
     from triton.runtime.autotuner import Autotuner
 
+    # Replace tensor pointers with Proxy object. Do not replace the other
+    # inputs since tl.constexprs can be inspected by the kernel.
+    proxy_kwargs = {
+        key: (Proxy() if isinstance(value, Tensor) else value)
+        for key, value in kwargs.items()
+    }
+
     if isinstance(kernel, Autotuner):
+        if len(kernel.configs) > 0:
+            # If we are autotuning, then it doesn't matter which version gets
+            # picked for tracing purposes, so lets pick the first one
+            proxy_kwargs = {**proxy_kwargs, **kernel.configs[0].kwargs}
         kernel = kernel.fn
 
-    infos = {name: MutationInfo() for name in tensors}
-    tracker = MutationTracker(infos)
-    tracker.visit(kernel.parse())
-    return [
-        name for name, info in infos.items() if info.mutated or info.used_in_unknown
-    ]
+    default_impls: Dict[str, FunctionType] = {}
+    mutation_ops = {
+        "store",
+        "atomic_add",
+        "atomic_cas",
+        "atomic_max",
+        "atomic_min",
+        "atomic_xchg",
+    }
+    try:
+        # Monkey patch all triton language functions
+        for name, impl in inspect.getmembers(tl, inspect.isfunction):
+            default_impls[name] = impl
+
+            if name in mutation_ops:
+                # Ops that mutate the first argument tensor pointer
+
+                def fn(pointer, *args, **kwargs):
+                    assert isinstance(pointer, Proxy)
+                    pointer.mutated = True
+                    return Scalar()
+
+            elif name == "make_block_ptr":
+                # Creates indirection to base tensor pointer
+
+                def fn(base, *args, **kwargs):  # type: ignore[misc] # different conditional signatures
+                    assert isinstance(base, Proxy)
+                    return base
+
+            elif name == "inline_asm_elementwise":
+                # If there's inline asm in the kernel, anything can happen.
+                # Do not optimize
+
+                def fn(*args, **kwargs):  # type: ignore[misc] # different conditional signatures
+                    raise IdentifyMutationException("inline_asm_elementwise in kernel")
+
+            else:
+                # Any operation not listed above, do not mutate a kernel
+                # pointer. They either return some sort of scalar or
+                # for the purposes of tracing cause an uninteresting
+                # side effect.
+
+                def fn(*args, **kwargs):  # type: ignore[misc] # different conditional signatures
+                    return Scalar()
```

Comment:
What about something like
```python
ptr = tl.where(cond, ptr1, ptr2)
tl.store(ptr ...
```

I think in general it would be safer to assume that functions propagate any pointers taken as argument, and have an exception for `tl.load`.

Reply:
Correct me If I am wrong but `tl.where` does not operate over tensor pointers, but rather tensor values (https://triton-lang.org/main/python-api/generated/triton.language.where.html)

Reply:
Essentially everything in triton is a "tensor", for example `ptr1` is a scalar tensor and `ptr1 + tl.arange(0, BLOCK)` is a rank 1 tensor of pointers.

Reply:
Okay, this makes sense. I'll update the algo, thanks, nice catch!

Reply:
That makes sense. Just tested on a simple case. I wonder whether tl.where is the only special operation or do we need to pessimize all analysis because a function could take two pointers and only read one propagate the other.

Reply:
The part about dynamically branching based on a tensor value pretty much makes static analysis be very pessimistic even if we were to analyze TTIR the original way I suggested.

Reply:
> The part about dynamically branching based on a tensor value pretty much makes static analysis be very pessimistic even if we were to analyze TTIR the original way I suggested.

Theoretically, I agree. But I'd suggest looking through a few internal Triton kernels to get a practical sense. My hunch is that, for the majority (if not all) internal Triton kernels, the static analysis should yield good results.